### DISCUSSION TITLE
Character vs Character contacts

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#105] [Q&A] [answered]
If I understand correctly, then whether a character collides with something is decided by the `character.ExtendedUpdate` and the broadphase/object filters that we give:

https://github.com/jrouwe/JoltPhysics.js/blob/d8339196c5f461c0ec8697d216ecec92c1e706f1/Examples/character_virtual.html#L177-L184

The example is using a moving layer:
https://github.com/jrouwe/JoltPhysics.js/blob/d8339196c5f461c0ec8697d216ecec92c1e706f1/Examples/character_virtual.html#L79-L80

Which I would assume enable character vs character collisions, as both would act as if they are "normal" dynamic objects. However, it doesn't seem to be the case? Looks like the characters cannot see each other. Am I missing something?

**Comment by jrouwe** [comment, marked as answer]
Hello,

Character vs character collisions are not supported with CharacterVirtual as described here:

https://jrouwe.github.io/JoltPhysics/index.html#character-controllers

You need to pair it with a rigid body (for which I still have to create a helper class).

**Reply by LeXXik** [reply]
Ah, I understand. Thank you! I will look into pairing it with a regular body then.

**Reply by LeXXik** [reply]
Trying to pair with a dynamic body and zero gravity factor. According to the docs:
```
Ensure that the character only collides with dynamic objects in this case.
```

That is to allow characters to hit dynamic bodies of other characters. However it will make the character hit its "own" dynamic body as well. Should I use subgroups for characters to filter out own paired dynamic bodies?

**Reply by LeXXik** [reply]
Hmm, maybe I should use a body filter in the extended update for that.

**Reply by LeXXik** [reply]
I ended up with creating a JS variant of the body filter and then using it during the extended update of the character. Maybe subgroups are more efficient? Anyhow, posting here this variant for reference. This will make character virtual ignore paired dynamic bodies:

```js
// create JS body filter
const bodyFilter = new Jolt.BodyFilterJS();

// implement its ShouldCollide and ShouldCollideLocked methods
bodyFilter.ShouldCollide = (inBodyID) => {
    if (pairedBody.GetID().GetIndexAndSequenceNumber() === Jolt.wrapPointer(inBodyID, Jolt.BodyID).GetIndexAndSequenceNumber()) {
        return false;
    }
    return true;
};
bodyFilter.ShouldCollideLocked = (inBody) => {
    return true;
};

// during the extended update use this filter, instead of the default one:
char.ExtendedUpdate(
    fixedStep,
    char.GetUp(),
    updateSettings,
    movingBPFilter,
    movingLayerFilter,
    bodyFilter , //  <---
    shapeFilter,
    allocator
);
```

**Reply by jrouwe** [reply]
I think a body filter should do just fine. Note that if the collision is rejected in ShouldCollide, you don't need to reject it again in ShouldCollideLocked (this is always called right after ShouldCollide).

**Reply by LeXXik** [reply]
Thank you! I have adjusted the sample code.

**Reply by dxu** [reply]
@LeXXik, this might be a result of my inexperience with wasm, but when should one use the "JS" implementation, and how did you know to look for the BodyFilterJS function as an option for implementing the body filter (e.g, as opposed to just using Jolt.BodyFilter)? was it just through browsing the idl?

**Reply by jrouwe** [reply]
All overridable C++ interfaces need a JS wrapper when using webidl. By convention, I've postfixed all these classes with JS.

**Comment by LeXXik** [comment]
Hmm, this stopped working for me for some reason. The paired dynamic body started to collide with the character shape. I know that there was a change regarding the character colliding with sensors, but not sure if that affected it. At least I noticed that my body filter `ShouldCollide` receives some ID that I don't recognize:
 - all other objects I have created start with 215xxxx, including the character
- the ID I get inside the callback starts with 124xxxx, which I assume should be that of the character, but it is not, so the filter fails to ignore the collision

![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/8e4caf80-8003-4ca7-a71f-81a2809d421b)

**Reply by jrouwe** [reply]
What caused this? A new version upgrade? And if so from which to which? And do you perhaps have a repro?

**Reply by LeXXik** [reply]
It was working when this discussion took place, so I suppose `0.17`. Stopped working after upgrading. Made a gist, just copy paste to your examples:

https://gist.github.com/LeXXik/254b9302fd6a080db70ef779badca0cc

In the example, I am dropping a dynamic sphere on a character. I want to make that sphere as a paired body, so the filter should ignore it.

**Reply by jrouwe** [reply]
I don't see how this could have worked before, the correct code is:

```
bodyFilter.ShouldCollide = (inBodyID) => {
	if (pairedBody.GetID().GetIndexAndSequenceNumber() === Jolt.wrapPointer(inBodyID, Jolt.BodyID).GetIndexAndSequenceNumber()) {
		return false;
	}
	return true;
};
```

your version:

```
bodyFilter.ShouldCollide = (inBodyID) => {
	if (pairedBody.GetID() === Jolt.wrapPointer(inBodyID, Jolt.BodyID)) {
		return false;
	}
	return true;
};
```

compares two BodyID objects, but these C++ classes don't get their equality operators exposed to JS so you're probably effectively comparing the address of a BodyID object on the stack and one on the heap which means they will never be the same.

**Reply by LeXXik** [reply]
Thank you! I am also not sure how it was working previously. Well, case closed. Thanks! I will modify the earlier snippet.
